lib/rucio/core/replica.py

```diff
@@ -2018,6 +2022,146 @@ def list_and_mark_unlocked_replicas(limit, bytes_=None, rse_id=None, delay_secon
     """
     List RSE File replicas with no locks.
 
+    :param limit:                    Number of replicas returned.
+    :param bytes_:                   The amount of needed bytes.
+    :param rse_id:                   The rse_id.
+    :param delay_seconds:            The delay to query replicas in BEING_DELETED state
+    :param only_delete_obsolete      If set to True, will only return the replicas with EPOCH tombstone
+    :param session:                  The database session in use.
+
+    :returns: a list of dictionary replica.
+    """
+
+    needed_space = bytes_
+    total_bytes = 0
+    rows = []
+
+    columns = [
+        Column("scope", InternalScopeString(get_schema_value('SCOPE_LENGTH'))),
+        Column("name", String(get_schema_value('NAME_LENGTH'))),
+    ]
+    temp_table_cls = create_temp_table(
+        "list_and_mark_unlocked_replicas",
+        *columns,
+        primary_key=columns,
+        session=session,
+    )
+
+    replicas_alias = aliased(models.RSEFileAssociation, name='replicas_alias')
+
+    stmt = select(
+        models.RSEFileAssociation.scope,
+        models.RSEFileAssociation.name,
+    ).with_hint(
+        models.RSEFileAssociation, "INDEX_RS_ASC(replicas REPLICAS_TOMBSTONE_IDX)  NO_INDEX_FFS(replicas REPLICAS_TOMBSTONE_IDX)", 'oracle'
+    ).where(
+        models.RSEFileAssociation.lock_cnt == 0,
+        models.RSEFileAssociation.rse_id == rse_id,
+        models.RSEFileAssociation.tombstone == OBSOLETE if only_delete_obsolete else models.RSEFileAssociation.tombstone < datetime.utcnow(),
+    ).where(
+        or_(models.RSEFileAssociation.state.in_((ReplicaState.AVAILABLE, ReplicaState.UNAVAILABLE, ReplicaState.BAD)),
+            and_(models.RSEFileAssociation.state == ReplicaState.BEING_DELETED, models.RSEFileAssociation.updated_at < datetime.utcnow() - timedelta(seconds=delay_seconds)))
+    ).outerjoin(
+        models.Source,
+        and_(models.RSEFileAssociation.scope == models.Source.scope,
+             models.RSEFileAssociation.name == models.Source.name,
+             models.RSEFileAssociation.rse_id == models.Source.rse_id)
+    ).where(
+        models.Source.scope.is_(None)  # Only try to delete replicas if they are not used as sources in any transfers
```

Comment:
This is what we had in the past, but had to relax it because it can completely block the deletion of some files, e.g. : you have a file on RSE A and RSE B and a request to transfer to RSE C that is unavailable. In that case there is a row in the source table for RSE A and B, and there is no way to delete at least one of these copies until RSE C is back online.

Reply:
I'm not sure what you mean by "relax". To my understanding this is just literal translation of the current behavior : https://github.com/rucio/rucio/blob/e7f1f44e00f44349a146b8e9e93ab942996896da/lib/rucio/core/replica.py#L2044 

Did I make a mistake and changed the current behavior in some way I didn't realize?

Reply:
This check was re-introduced by us last year, we did have to relax it before, that's what Cedric mentions.
But I think it is fine now, this PR does not change that behaviour.